### PR TITLE
DQN implementation

### LINKED ISSUE #69: Add DQN algorithms

### PR DESCRIPTION
**Pull request by axitkhurana**
Hi there,

Thanks for writing this library! I wrote a simple DQN algorithm using other parts of the library and reading the [DQN paper](https://www.cs.toronto.edu/~vmnih/docs/dqn.pdf). Didn't use the convolutional layers as mentioned in the paper or the preprocessing steps

Did test on CartPole-v1, did better than I expected but not sure if that's good enough.

Would love for the maintainers to take a look and see if I can add or improve something here.

Fixes #69 

Thanks!

### CODE REVIEW

**Comment by jachiam** [comment]
Hi @axitkhurana! This is great code quality for a DQN implementation and it comes quite close to meeting the bar for Spinning Up. There are a few things that would need to be different, though, and then there will be two major delays before I could merge it into master.


**Necessary changes:**

- Target networks appear absent from this implementation, but they are fairly crucial for DQN. Please include them and follow the original DQN convention, of periodically copying the target network parameters from the main network (as opposed to the DDPG/TD3/SAC style of polyak-averaging the target networks).

- While saving all updates for the end of each episode works reasonably well for the environments tested with TD3, the right thing to do for DQN would be to have one update per environment interaction step.

**Delays to expect:**

A rule I'm following for Spinning Up implementations is that there will be no algorithms in the code base which don't have corresponding explanation pages on the website. This is something that takes a lot of thinking, writing, and editing on my end, and I'm bottlenecked on writing additional content until probably around April. This is the largest delay that will prevent a near-term merge to master.

The other delay is that Spinning Up implementations are required to have benchmarks that demontrate reasonable performance. This is mostly a matter of time and compute (and possibly hyperparameter tuning), which you're free to spend if you have it, but otherwise I imagine I'll need to do it---and it will probably be a while before I can spare the time for it. 

But---those caveats on merging aside---kudos for the very nice work! Thanks for contributing!

**Comment by axitkhurana** [comment]
Thanks for looking at it @jachiam!

> Target networks appear absent from this implementation, but they are fairly crucial for DQN. Please include them and follow the original DQN convention, of periodically copying the target network parameters from the main network (as opposed to the DDPG/TD3/SAC style of polyak-averaging the target networks).

* Will add the target networks, didn't see one in the first paper.

> While saving all updates for the end of each episode works reasonably well for the environments tested with TD3, the right thing to do for DQN would be to have one update per environment interaction step.

* Sorry about the misleading [comment](https://github.com/openai/spinningup/pull/95/files#diff-5e6da5db836e01e9a3a12ac485b86804R209), actual [update](https://github.com/openai/spinningup/pull/95/files#diff-5e6da5db836e01e9a3a12ac485b86804R227) happens on each interaction after `t > start_steps`

* I'll also add draft explanation pages based on the paper and current documentation and see if I can add some benchmarks.

Appreciate your time!

**Comment by axitkhurana** [comment]
Added target network and now it works perfectly on `CartPole-v1`:

```
Episode 86       EpRet 500.000   EpLen 500
Episode 87       EpRet 500.000   EpLen 500
Episode 88       EpRet 500.000   EpLen 500
Episode 89       EpRet 500.000   EpLen 500
Episode 90       EpRet 500.000   EpLen 500
Episode 91       EpRet 500.000   EpLen 500
Episode 92       EpRet 500.000   EpLen 500
Episode 93       EpRet 500.000   EpLen 500
Episode 94       EpRet 500.000   EpLen 500
Episode 95       EpRet 500.000   EpLen 500
Episode 96       EpRet 500.000   EpLen 500
Episode 97       EpRet 500.000   EpLen 500
Episode 98       EpRet 500.000   EpLen 500
Episode 99       EpRet 500.000   EpLen 500
-------------------------------------
|    AverageEpRet |             500 |
|        StdEpRet |               0 |
|        MaxEpRet |             500 |
|        MinEpRet |             500 |
|           EpLen |             500 |
```

**Comment by karanchahal** [comment]
Hey @axitkhurana ! Great work. I would also like to help you benchmark this thing. We can also include Rainbow DQN features such as a Prioritized Experience Replay, Dueling DQN's to further stabilise the performance as vanilla DQN is known to be a bit shaky. Adding the Dueling DQN is pretty trivial as we already have the target and policy network.

@jachiam Can you guide us on what envs should to benchmark on ? I have the compute to do so, hence I would be glad to help :)

**Comment by karanchahal** [comment]
I have coded a Double DQN solution [here](https://colab.research.google.com/drive/1gDPogAlbcMGNs5NAfuF_5C522ZdzWfk7).

It trains Cartpole as expected (comes to 200 maximum reward for 200 consecutive epochs). I've observed that it performs slightly better than vanilla DQN but no way to be sure because RL is so misleading.

Do you want together on getting DQN incorporated in SpinningUp @axitkhurana  ? I have a CPU system with 40 cores in my lab so training should not be a problem :)

**Comment by karanchahal** [comment]
Also just wanted to say to @jachiam that your MPI code is beautiful ! So clean and just.. right. Can you point us to a resource from where you learnt modular efficient coding like this ?

**Comment by axitkhurana** [comment]
Hi @karanchahal,

Happy to work on benchmarking and should have the resources needed with the EC2 instances. Last I had a chance to talk to Josh, he was planning to write up docs similar to VPG and decide Envs for benchmarking.

I did work on 

* dueling-dqn: https://github.com/axitkhurana/spinningup/tree/dueling-dqn and
* double-dqn https://github.com/axitkhurana/spinningup/tree/double-dqn (this one has a small bug, @karanchahal if you have time, let me know if you can spot anything).

in the mean time. Planning to implement some other papers on the list in April.

Let me know if we can do anything here @jachiam, appreciate your help at the Spinning Up workshop! :)
